src/plugins/discover/public/application/components/discover_grid/discover_grid.tsx

```diff
@@ -0,0 +1,329 @@
+  onSetColumns: (columns: string[]) => void;
+  /**
+   * function to change sorting of the documents
+   */
+  onSort: (sort: string[][]) => void;
+  /**
+   * Array of documents provided by Elasticsearch
+   */
+  rows?: ElasticSearchHit[];
+  /**
+   * The max size of the documents returned by Elasticsearch
+   */
+  sampleSize: number;
+  /**
+   * Grid display settings persisted in Elasticsearch (e.g. column width)
+   */
+  settings?: DiscoverGridSettings;
+  /**
+   * Saved search description
+   */
+  searchDescription?: string;
+  /**
+   * Saved search title
+   */
+  searchTitle?: string;
+  /**
+   * Discover plugin services
+   */
+  services: DiscoverServices;
+  /**
+   * Determines whether the time columns should be displayed (legacy settings)
+   */
+  showTimeCol: boolean;
+  /**
+   * Current sort setting
+   */
+  sort: SortPairArr[];
+}
+
+export const EuiDataGridMemoized = React.memo((props: EuiDataGridProps) => {
+  return <EuiDataGrid {...props} />;
+});
+
+export const DiscoverGrid = ({
+  ariaLabelledBy,
+  columns,
+  defaultColumns,
+  indexPattern,
+  onAddColumn,
+  onFilter,
+  onRemoveColumn,
+  onResize,
+  onSetColumns,
+  onSort,
+  rows,
+  sampleSize,
+  searchDescription,
+  searchTitle,
+  services,
+  settings,
+  showTimeCol,
+  sort,
+}: DiscoverGridProps) => {
+  const [expanded, setExpanded] = useState<ElasticSearchHit | undefined>(undefined);
+
+  /**
+   * Pagination
+   */
+  const [pagination, setPagination] = useState({ pageIndex: 0, pageSize: defaultPageSize });
+
+  const rowsLength = rows ? rows.length : 0;
+  const pageCount = Math.ceil(rowsLength / pagination.pageSize);
+  const isOnLastPage = pagination.pageIndex === pageCount - 1;
+
+  const paginationObj = useMemo(() => {
+    const onChangeItemsPerPage = (pageSize: number) =>
+      setPagination((paginationData) => ({ ...paginationData, pageSize }));
+
+    const onChangePage = (pageIndex: number) =>
+      setPagination((paginationData) => ({ ...paginationData, pageIndex }));
+
+    return {
+      onChangeItemsPerPage,
+      onChangePage,
+      pageIndex: pagination.pageIndex > pageCount - 1 ? 0 : pagination.pageIndex,
+      pageSize: pagination.pageSize,
+      pageSizeOptions: pageSizeArr,
+    };
+  }, [pagination, pageCount]);
+
+  /**
+   * Sorting
+   */
+  const sortingColumns = useMemo(() => sort.map(([id, direction]) => ({ id, direction })), [sort]);
+
+  const onTableSort = useCallback(
+    (sortingColumnsData) => {
+      onSort(sortingColumnsData.map(({ id, direction }: SortObj) => [id, direction]));
+    },
+    [onSort]
+  );
+
+  /**
+   * Cell rendering
+   */
+  const renderCellValue = useMemo(() => getRenderCellValueFn(indexPattern, rows), [
+    rows,
+    indexPattern,
+  ]);
+
+  /**
+   * Render variables
+   */
+  const showDisclaimer = rowsLength === sampleSize && isOnLastPage;
+  const randomId = useMemo(() => htmlIdGenerator()(), []);
+
+  const rowCount = useMemo(() => (rows ? rows.length : 0), [rows]);
```

Comment:
Do we need both rowCount and rowsLength?

Reply:
no, this is redundant, good catch, improved that 👍